src/nvim/terminal.c

```diff
@@ -413,7 +423,16 @@ void terminal_enter(bool process_deferred)
   mapped_ctrl_c = save_mapped_ctrl_c;
   unshowmode(true);
   redraw(term->buf != curbuf);
+  if (curbuf == term->buf) {
+    curwin->w_p_fen = save_w_p_fen;
+    if (mouse_event == false) {
+      // Upon leaving Terminal mode, keep cursor where Terminal showed it
+      curwin->w_cursor.lnum = row_to_linenr(term, term->cursor.row);
+      curwin->w_cursor.col = term->cursor.col;
+    }
+  }
```

Comment:
This should be broken out into a separate PR, it's a rather high-priority improvement.